sdk/src/extras/disk.rs

```diff
@@ -0,0 +1,711 @@
+            })
+            .send()
+            .await
+            .context("finalizing the disk failed")?;
+
+        if self.image_info.is_some() {
+            self.create_image().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn create_image(&self) -> Result<()> {
+        let Some(image_info) = &self.image_info else {
+            bail!("no snapshot information provided");
+        };
+
+        // At this point, unwinding is not as important as before. the user
+        // has uploaded their file to a disk and finalized that disk, making
+        // a snapshot out of it. if this step fails, they can always
+        // manually make an image out of the snapshot later and be sure that
+        // the snapshot's contents are the same.
+        let snapshot_id = self
+            .client
+            .snapshot_view()
+            .project(&self.project)
+            .snapshot(NameOrId::Name(image_info.snapshot.clone()))
+            .send()
+            .await?;
+
+        self.client
+            .image_create()
+            .project(&self.project)
+            .body(ImageCreate {
+                name: image_info.image.clone(),
+                description: image_info.image_description.clone(),
+                os: image_info.image_os.clone(),
+                version: image_info.image_version.clone(),
+                source: ImageSource::Snapshot(snapshot_id.id),
+            })
+            .send()
+            .await?;
+
+        Ok(())
+    }
+
+    async fn cleanup(&self) -> Result<()> {
+        if self
+            .cleanup_started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .is_err()
+        {
+            return Ok(());
+        }
+
+        let disk_state = match self.wait_for_disk().await {
+            Ok(state) => state,
+            Err(e) => match e {
+                // No disk, no problem.
+                Error::ErrorResponse(rv) if rv.status() == 404 => return Ok(()),
+                _ => Err(e)?,
+            },
+        };
+
+        // Disk is not partially imported, no cleanup needed.
+        if !matches!(
+            disk_state,
+            DiskState::ImportReady
+                | DiskState::ImportingFromBulkWrites
+                | DiskState::ImportingFromUrl
+                | DiskState::Finalizing
+        ) {
+            return Ok(());
+        }
+
+        if matches!(disk_state, DiskState::ImportingFromBulkWrites) {
+            self.unwind_disk_bulk_write_stop().await?;
+        }
+
+        self.unwind_disk_finalize().await?;
+        self.unwind_disk_delete().await
+    }
+
+    async fn check_for_existing_disk(&self) -> Result<()> {
+        err_if_object_exists(
+            format!("disk \"{}\" exists already", &*self.disk),
+            self.client
+                .disk_view()
+                .project(&self.project)
+                .disk(NameOrId::Name(self.disk.clone()))
```

Comment:
Same as above.

Reply:
Fixed in 7cc0419cc159d5b4a48a8d8c9bec1dddc7a53cb3